superchain/superchain.go

```diff
@@ -76,55 +78,57 @@ const (
 )
 
 type ChainConfig struct {
-	Name         string `yaml:"name"`
-	ChainID      uint64 `yaml:"chain_id"`
-	PublicRPC    string `yaml:"public_rpc"`
-	SequencerRPC string `yaml:"sequencer_rpc"`
-	Explorer     string `yaml:"explorer"`
+	Name         string `yaml:"name" toml:"name"`
+	ChainID      uint64 `yaml:"chain_id" toml:"chain_id"`
+	PublicRPC    string `yaml:"public_rpc" toml:"public_rpc"`
+	SequencerRPC string `yaml:"sequencer_rpc" toml:"sequencer_rpc"`
+	Explorer     string `yaml:"explorer" toml:"explorer"`
 
-	SuperchainLevel SuperchainLevel `yaml:"superchain_level"`
+	SuperchainLevel SuperchainLevel `yaml:"superchain_level" toml:"superchain_level"`
 
 	// If StandardChainCandidate is true, standard chain validation checks will
 	// run on this chain even if it is a frontier chain.
-	StandardChainCandidate bool `yaml:"standard_chain_candidate,omitempty"`
+	StandardChainCandidate bool `yaml:"standard_chain_candidate,omitempty" toml:"standard_chain_candidate,omitempty"`
 
 	// If SuperchainTime is set, hardforks times after SuperchainTime
 	// will be inherited from the superchain-wide config.
-	SuperchainTime *uint64 `yaml:"superchain_time"`
+	SuperchainTime *uint64 `yaml:"superchain_time" toml:"superchain_time"`
 
-	BatchInboxAddr Address `yaml:"batch_inbox_addr"`
-
-	Genesis ChainGenesis `yaml:"genesis"`
+	BatchInboxAddr Address `yaml:"batch_inbox_addr" toml:"batch_inbox_addr"`
 
 	// Superchain is a simple string to identify the superchain.
 	// This is implied by directory structure, and not encoded in the config file itself.
-	Superchain string `yaml:"-"`
+	Superchain string `yaml:"-" toml:"-"`
 	// Chain is a simple string to identify the chain, within its superchain context.
 	// This matches the resource filename, it is not encoded in the config file itself.
-	Chain string `yaml:"-"`
+	Chain string `yaml:"-" toml:"-"`
 
 	// Hardfork Configuration Overrides
-	HardForkConfiguration `yaml:",inline"`
+	HardForkConfiguration `yaml:",inline" toml:",inline"`
+
+	BlockTime           uint64 `yaml:"block_time" toml:"block_time"`
+	SequencerWindowSize uint64 `yaml:"seq_window_size" toml:"seq_window_size"`
+
+	Genesis ChainGenesis `yaml:"genesis" toml:"genesis"`
 
-	BlockTime           uint64 `yaml:"block_time"`
-	SequencerWindowSize uint64 `yaml:"seq_window_size"`
+	Addresses AddressList `toml:"addresses"`
```

Comment:
I suppose this data is now available both at the existing top level `Addresses` mapping _and_ at `config.Addresses`. We could probably "hide it" by not exporting it, but this would require a bit of extra work to get it to unmarshal properly. 

I think it could make sense though, to use the single unified input TOML file but umarshal it three times. Once into an unmodified `ChainConfig` (i.e. it does not have Addresses field), and then into the top level `Addresses` field and again into the top level `GenesisSystemConfigs` field. If we did that, we wouldn't be changing the API of the Go bindings at all.

I don't feel strongly though, I think longer term a single mapping for a single set of TOML files makes the most sense, and we should just go ahead and make the breaking change and tidy it up when we find the time. For now, this is probably an acceptable bit of duplication.

Reply:
Agreed that the breaking change will be the best solution. Issue filed here to capture that work: https://github.com/ethereum-optimism/superchain-registry/issues/391